crates/debugging/src/trace/function/trace.rs

```diff
@@ -0,0 +1,131 @@
+        if contracts_data_store.is_fork(&class_hash) {
+            return Err(FunctionTraceError::ForkContract);
+        }
+        let program_artifact = contracts_data_store
+            .get_program_artifact(&class_hash)
+            .expect("program artifact should be present for local contracts");
+
+        let program = &program_artifact.program;
+
+        let sierra_program_registry = ProgramRegistry::<CoreType, CoreLibfunc>::new(program)
+            .expect("failed to create Sierra program registry");
+
+        let mut call_stack = CallStack::new();
+        let mut stacks = Vec::new();
+
+        for statement_idx in get_sierra_statements(class_hash, call_trace, contracts_data_store) {
+            let function_name = FunctionName::from_program(statement_idx, program);
+
+            let stack = call_stack.new_stack(function_name);
+
+            stacks.push(stack.clone());
+
+            let statement = program
+                .statements
+                .get(statement_idx.0)
+                .expect("statement should be present");
+
+            match statement {
+                GenStatement::Invocation(invocation) => {
+                    let libfunc = sierra_program_registry.get_libfunc(&invocation.libfunc_id);
+
+                    if let Ok(CoreConcreteLibfunc::FunctionCall(_)) = &libfunc {
+                        call_stack.enter_function_call(stack);
+                    }
+                }
+                GenStatement::Return(_) => {
+                    call_stack.exit_function_call();
+                }
+            }
+        }
+
+        Ok(build_function_trace(stacks))
+    }
+}
+
+/// Retrieves vector of [`StatementIdx`] from the given [`CallTrace`].
+fn get_sierra_statements(
+    class_hash: ClassHash,
+    call_trace: &CallTrace,
+    contracts_data_store: &ContractsDataStore,
+) -> Vec<StatementIdx> {
+    let casm_debug_info = contracts_data_store
+        .get_casm_debug_info(&class_hash)
+        .expect("Cairo program debug info should be present");
+
+    let casm_level_info = build_casm_level_info(call_trace);
+
+    map_pcs_to_sierra_statement_ids(casm_debug_info, &casm_level_info)
+        .into_iter()
+        .filter_map(Option::from)
+        .collect()
+}
+
+/// Builds a [`CasmLevelInfo`] from the given [`CallTrace`].
+fn build_casm_level_info(call_trace: &CallTrace) -> CasmLevelInfo {
+    let vm_trace = call_trace
+        .vm_trace
+        .as_ref()
+        .expect("vm trace should be present")
+        .iter()
+        .map(|value| TraceEntry {
+            pc: value.pc,
+            ap: value.ap,
+            fp: value.fp,
+        })
+        .collect();
+
+    CasmLevelInfo {
+        run_with_call_header: false,
+        vm_trace,
+    }
+}
+
+/// Builds a [`FunctionTrace`] from the given stacks of function names.
+fn build_function_trace(stacks: Vec<Vec<FunctionName>>) -> FunctionTrace {
+    let mut dummy_root = FunctionNode::new(FunctionName::NonInlined("root".to_string()));
```

Comment:
nit: `dummy_root` works, but `placeholder_root` might better convey that it's a temporary node for building the tree:
```suggestion
    let mut placeholder_root = FunctionNode::new(FunctionName::NonInlined("root".to_string()));
```